File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
   static bool is_sorted(vector<int>&nums){
        for(int i=0;i<nums.size()-1;i++){
            if(nums[i]<nums[i+1]){
                return false;
            }
        }
       return true;
    }
    void nextPermutation(vector<int>& nums) {
        if(is_sorted(nums)){
            sort(nums.begin(),nums.end());
            cout<<"yes"<<endl;
            return;
        }
        int firstSwap , secondSwap;
       
        for(int i=nums.size()-1;i>=1;i--){
            if(nums[i]> nums[i-1]){
                firstSwap = i-1;
               
                break;
            }
        }
        
        for(int i=nums.size()-1;i>=0;i--){
            if(nums[i]>nums[firstSwap]){
                secondSwap =  i;
              
                break;
            }
        }
        cout<<firstSwap<<" "<<secondSwap<<endl;
        swap(nums[firstSwap],nums[secondSwap]);
        if(firstSwap+2< nums.size()){
            reverse(nums.begin()+firstSwap+1, nums.end());
        }
        
    }
};

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int first =  nums.length-1;
        while(first>=1 && nums[first-1] >= nums[first]){
            first--;
        }
        
        if(first==0){
            Arrays.sort(nums);
            return;
        }
        first--;
        int second =  nums.length-1;
        while(second> first && nums[second] <= nums[first]){
            second--;
        }
       
        swap(nums,first,second);
        reverse(nums,first+1);
        
    }
    
    public static void swap(int[] nums, int i , int j){
        int temp = nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
    
    public static void reverse(int[] nums, int i){
        int j=nums.length-1;
        while(i<j){
            swap(nums,i,j);
            i++;
            j--;
            
        }
        return;
    }
}